spec/CRM/StateMachineSpec.hs

```diff
@@ -0,0 +1,68 @@
+
+spec :: Spec
+spec =
+  describe "StateMachine" $ do
+    describe "run" $ do
+      describe "switch machine" $ do
+        it "outputs () when the state is False" $ do
+          run (Basic $ switchMachine SFalse) () `shouldOutput` ()
+
+        it "outputs () when the state is True" $ do
+          run (Basic $ switchMachine STrue) () `shouldOutput` ()
+
+      describe "lock door machine" $ do
+        it "outputs Opened when it is closed and receive an Open command" $ do
+          run (Basic $ lockDoorMachine SIsLockClosed) LockOpen `shouldOutput` LockOpened
+
+        it "outputs NoOp when it is closed and receive a Close command" $ do
+          run (Basic $ lockDoorMachine SIsLockClosed) LockClose `shouldOutput` LockNoOp
+
+        it "outputs Locked when it is closed and receive a Lock command" $ do
+          run (Basic $ lockDoorMachine SIsLockClosed) LockLock `shouldOutput` LockLocked
+
+        it "outputs NoOp when it is closed and receive an Unlock command" $ do
+          run (Basic $ lockDoorMachine SIsLockClosed) LockUnlock `shouldOutput` LockNoOp
+
+        it "outputs NoOp when it is open and receive an Open command" $ do
+          run (Basic $ lockDoorMachine SIsLockOpen) LockOpen `shouldOutput` LockNoOp
+
+        it "outputs Closed when it is open and receive a Close command" $ do
+          run (Basic $ lockDoorMachine SIsLockOpen) LockClose `shouldOutput` LockClosed
+
+        it "outputs NoOp when it is open and receive a Lock command" $ do
+          run (Basic $ lockDoorMachine SIsLockOpen) LockLock `shouldOutput` LockNoOp
+
+        it "outputs NoOp when it is open and receive an Unlock command" $ do
+          run (Basic $ lockDoorMachine SIsLockOpen) LockUnlock `shouldOutput` LockNoOp
+
+        it "outputs NoOp when it is locked and receive an Open command" $ do
+          run (Basic $ lockDoorMachine SIsLockLocked) LockOpen `shouldOutput` LockNoOp
+
+        it "outputs NoOp when it is locked and receive a Close command" $ do
+          run (Basic $ lockDoorMachine SIsLockLocked) LockClose `shouldOutput` LockNoOp
+
+        it "outputs NoOp when it is locked and receive a Lock command" $ do
+          run (Basic $ lockDoorMachine SIsLockLocked) LockLock `shouldOutput` LockNoOp
+
+        it "outputs Unlocked when it is locked and receive an Unlock command" $ do
+          run (Basic $ lockDoorMachine SIsLockLocked) LockUnlock `shouldOutput` LockUnlocked
+
+        it "outputs Locked when it is open and receive a Close and a Lock command" $ do
+          let
+            runOnce = snd $ run (Basic $ lockDoorMachine SIsLockOpen) LockClose
+          run runOnce LockLock `shouldOutput` LockLocked
+
+        it "outputs Opened when it is locked and receive a Unlock and an Open command" $ do
+          let
+            runOnce = snd $ run (Basic $ lockDoorMachine SIsLockLocked) LockUnlock
+          run runOnce LockOpen `shouldOutput` LockOpened
```

Comment:
It would be neat to have a function of the form:

```haskell
initialState :: StateMachine a b -> InitialState state
initialState (Basic a) = BaseMachine.initialState a
initialState (Compose a _) = initialState a
```

This way we could test both the output and the new state of the machine. It could also be useful to expose it to the library users outside of tests.

What do you think?

Reply:
I'd say that would definitely make sense.

I'll open an issue for it, mentioning that tests should be improved